### PR TITLE
Trigger NotFound404Controller if no Route was found

### PR DESCRIPTION
**Pull request by Chemaclass**
## 📚 Description

Inspired by the issue: https://github.com/gacela-project/router/issues/10 but instead of throwing an exception, simply change the header location to 404.

## 🔖 Changes

- Create a `NotFound404Controller`
- Create a new `Route` using the `NotFound404Controller` if no route was found
- Rename tests folder from Unit to Feature
- Create a parent TestCase for those mocking the `header()`
-

### CODE REVIEW

**Review by JesusValeraDev — APPROVED** [review]

File: tests/Unit/Router/RouterRedirectTest.php [hunk 1]
```diff
@@ -110,7 +110,13 @@ static function (Routes $routes) use ($destination, $statusCode, $method): void
             },
         );
 
-        self::assertNull($testHeaders);
+        self::assertSame([
```
Comment (antonio-gg-dev) [inline tests/Unit/Router/RouterRedirectTest.php]: For the next ErrorHandling feature I would move this test to a new class. There are also other tests such this one: https://github.com/gacela-project/router/blob/80dd2fa300fdcf01a1749aef3cb2d40baab5891b/tests/Unit/Router/RouterMatchTest.php#L29 that with this implementation is now not testing the desired behavior.

**Review by antonio-gg-dev — CHANGES_REQUESTED** [review]

File: tests/Feature/HeaderTestCase.php [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+namespace GacelaTest\Feature;
+
+use PHPUnit\Framework\TestCase;
+
+include_once __DIR__ . '/header.php';
+
+abstract class HeaderTestCase extends TestCase
+{
+    /**
+     * @runInSeparateProcess
```
Comment (antonio-gg-dev) [inline tests/Feature/HeaderTestCase.php:14]: This doesn't work if we use a trait right? In other words, we are skipping the composition over inheritance due to a technical limitation. @Chemaclass

File: tests/Feature/Router/ErrorHandlingTest.php [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+namespace GacelaTest\Feature\Router;
+
+use Gacela\Router\Entities\Request;
+use Gacela\Router\Router;
+use GacelaTest\Feature\HeaderTestCase;
+
+final class ErrorHandlingTest extends HeaderTestCase
+{
+    public function test_not_redirect_non_registered_method(): void
```
Comment (antonio-gg-dev) [inline tests/Feature/Router/ErrorHandlingTest.php]: I think this name could be improved, it is not correctly defining the situation under test, it should be test_respod_status_404_when_the_uri_does_not_matches.

File: tests/Feature/Router/RouterMatchTest.php [hunk 4]
```diff
@@ -26,18 +26,6 @@ public function test_respond_when_everything_matches(): void
...
-        $_SERVER['REQUEST_METHOD'] = 'GET';
-
-        $this->expectOutputString('');
-
-        Router::configure(static function (Routes $routes): void {
-            $routes->get('other/uri', FakeController::class, 'basicAction');
-        });
-    }
-
     public function test_not_respond_when_the_method_does_not_matches(): void
```
Comment (antonio-gg-dev) [inline tests/Feature/Router/RouterMatchTest.php]: You also forgot this test, I think you should move it to ErrorHandlingTest as well.

Comment (antonio-gg-dev) [reply tests/Feature/Router/RouterMatchTest.php]: And also the test called test_match_does_not_matches_other_methods

**Review by antonio-gg-dev — APPROVED** [review]
I really like the changes you have made @Chemaclass , although I have left you a couple of comments that I think you should fix, I approve it to speed up.

File: tests/Feature/HeaderTestCase.php (hunk 2 above)
Comment (Chemaclass) [reply tests/Feature/HeaderTestCase.php:14]: That's 100% correct. This works on a class level, but not on trait level

**Pull request merged**